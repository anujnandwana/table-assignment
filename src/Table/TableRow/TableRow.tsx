import React from "react";
import "../Table.css";

interface TableRowProps {
  rowData: any;
  tableHeadings: string[];
  isSelected: boolean;
  onSelectRow: (isSelected: boolean, selectedId: string) => void;
}

const TableRow: React.FC<TableRowProps> = ({
  rowData,
  tableHeadings,
  isSelected,
  onSelectRow,
}) => {
    
  /**
   * Handles the selection and deselection of table rows
   * @param event - The change event from the checkbox input
   * @param selectedData - The row data associated with the checkbox
   */
  const handleSelectRow = (
    event: React.ChangeEvent<HTMLInputElement>,
    selectedId: string
  ) => {
    onSelectRow(event.target.checked, selectedId);
  };

  /**
   * Handles key up events for the checkbox for accessibility
   * @param event - The keyboard event
   * @param rowData - The row data associated with the checkbox
   */
  const handleKeyUp = (
    event: React.KeyboardEvent<HTMLInputElement>,
    selectedId: string
  ) => {
    if (event.key === "Enter") {
      onSelectRow(!isSelected, selectedId);
    }
  };

  return (
    <tr className={isSelected ? "active " : ""}>
      {tableHeadings.map((heading, colIndex) => (
        <td
          key={`row-${rowData.id}-col-${colIndex}`}
          aria-label={heading}
          className={heading}
          data-status={heading === "status" ? rowData[heading] : undefined}
        >
          {heading === "id" ? (
            <input
              type="checkbox"
              className={"row-selector-" + rowData.id}
              onChange={(event) => handleSelectRow(event, rowData.id)}
              onKeyUp={(event) => handleKeyUp(event, rowData.id)}
              checked={isSelected}
              aria-checked={isSelected}
              aria-labelledby={`row-${rowData.id}-label`}
            />
          ) : (
            rowData[heading]
          )}
        </td>
      ))}
    </tr>
  );
};

export default TableRow;
