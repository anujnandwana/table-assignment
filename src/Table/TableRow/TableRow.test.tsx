import TableRow from './TableRow';
import { mount } from 'enzyme';
import { TableData } from '../../data/data';


describe('TableRow component', () => {
  const mockData: TableData[] = [
    { id: "mockedId", device: 'Device 1', path: '/path/1', name: 'jui', status: 'available' },
    { id: "mockedId1", device: 'Device 2', path: '/path/2', name: 'jui2', status: 'available' }

  ];

  const onSelectRow = jest.fn()
  const rowData = mockData[0];
  const tableHeadings = Object.keys(mockData[0]);

  test('TableRow component should match snapshot', () => {
    const component = mount(<TableRow onSelectRow={onSelectRow} tableHeadings={tableHeadings} isSelected={false} rowData = {rowData}/>);
    expect(component).toMatchSnapshot();
  });

  test('TableRow Checkbox on click select rows when checked and remove rows when unchecked', () => {
    const onSelectRow = jest.fn()
    const component = mount(<TableRow onSelectRow={onSelectRow} tableHeadings={tableHeadings} isSelected={false} rowData = {rowData}/>);
    const rowCheck = component.find("input");

    rowCheck.at(0).simulate('change', { target: { checked: true } });

    expect(rowCheck.length).toBe(1);
    expect(onSelectRow).toHaveBeenCalledWith(true, "mockedId")

    rowCheck.at(0).simulate('change', { target: { checked: false } });

    expect(onSelectRow).toHaveBeenLastCalledWith(false, "mockedId")
  });

  test('selectAll Checkbox on key Up select all rows when checked', () => {
    const onSelectRow = jest.fn()
    const component = mount(<TableRow onSelectRow={onSelectRow} tableHeadings={tableHeadings} isSelected={false} rowData = {rowData}/>);
    const rowCheck = component.find("input");

    rowCheck.at(0).simulate('keyup', { key: 'Enter', keyCode: 13, which: 13 });

    expect(rowCheck.length).toBe(1);
    expect(onSelectRow).toHaveBeenCalledWith(true, "mockedId");
  });
  
});
