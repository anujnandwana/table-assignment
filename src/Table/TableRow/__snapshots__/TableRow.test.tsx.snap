// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow component TableRow component should match snapshot 1`] = `ReactWrapper {}`;
